### PR TITLE
feat(scanner): add option not to follow symlinks

### PR DESCRIPTION
**Pull request by lutzky**
Never consider a symlink to be a subdirectory; this prevents hanging when the symlink points to a slow/inaccessible filesystem.

#### Description

Instead of using `is_dir`, which follows symlinks, use [`symlink_metadata`](https://doc.rust-lang.org/std/fs/fn.symlink_metadata.html), which doesn't.

#### Motivation and Context

If the current directory has a symlink to a slow filesystem (e.g. an inaccessible network filesystem), `is_dir` hangs indefinitely, disregarding `scan_timeout`.

This should help with some instances of #312.

#### Screenshots (if appropriate):

n/a

#### How Has This Been Tested?

- [ ] I have tested using **MacOS**
- [x] I have tested using **Linux**
- [ ] I have tested using **Windows**

I've tested by running `starship timings` with network disconnected in a directory which includes a symlink to a network filesystem. Without my change it hangs, with my change it doesn't.

#### Checklist:

- [x] I have updated the documentation accordingly
- [x] I have updated the tests accordingly

### CODE REVIEW

**Comment by davidkna** [comment]
Following symlinks does seem somewhat useful, I assume there is no option to ignore cross-device symlinks only?

**Comment by lutzky** [comment]
The trick is determining whether or not a symlink is cross-device *without* calling `stat` on the destination (because that's the bit that might fail). https://github.com/stemjail/mnt-rs seems to be able to do it, but doesn't appear very portable. WDYT?

**Comment by davidkna** [comment]
> The trick is determining whether or not a symlink is cross-device _without_ calling `stat` on the destination (because that's the bit that might fail). https://github.com/stemjail/mnt-rs seems to be able to do it, but doesn't appear very portable. WDYT?

Try using `systemstat` which we already depend on, relevant functions will likely return an error on unsupported operating systems which should be handled.

**Comment by lutzky** [comment]
I've tried out some stuff in https://github.com/starship/starship/pull/5325/commits/a99335a127a4b2369dffa60b829b4fe05fccbc14; WDYT? Reasonable or too brittle?

**Comment by lutzky** [comment]
There's an issue with `systemstat`: [`mounts`](https://docs.rs/systemstat/latest/x86_64-unknown-linux-gnu/src/systemstat/platform/linux.rs.html#641-654) proceeds to immediately `stat` each filesystem, causing it to wait on network. :disappointed: 

I guess one path forward is submitting a patch to `systemstat` to make a no-stat-just-the-files option... WDYT?

**Comment by davidkna** [comment]
> I guess one path forward is submitting a patch to `systemstat` to make a no-stat-just-the-files option... WDYT?

The crate is somewhat minimalistic, so I'm uncertain if the maintainers will be receptive to such a change.

**Comment by lutzky** [comment]
So, given these limitations, how about we go back to the "don't follow symlinks in this case" option? Perhaps as a configurable?

**Comment by davidkna** [comment]
I think making it configurable is fine, unless you want to try making the timeout itself more reliable.

**Comment by lutzky** [comment]
Making the timeout itself more reliable is more challenging (`D` is called "uninterruptable" for a reason, you'd end up accumulating stale threads, it might be worth doing but would be messy). For now, I've made it configurable. PTAL

**Comment by lutzky** [comment]
Finally added tests. I believe that the failures on nightly are unrelated to my changes. Do my changes seem OK now?

**Review by davidkna — APPROVED** [review]
LGTM

**Comment by lutzky** [comment]
Still blocked on https://github.com/starship/starship/pull/5467 for passing the checks.

**Comment by davidkna** [comment]
Thanks for the contribution @lutzky!

**Pull request merged**
